### PR TITLE
Add support for ``StatefulDataLoader``

### PR DESCRIPTION
**Pull request by joecummings**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

This PR adds support for the `StatefulDataLoader` class from the PyTorch library torchdata.

**FAQs**
1. **Why?** This is necessary for resuming from checkpoints mid-epoch - a feature set we will need to support for step-based checkpointing.
2. **Only full finetune single device?** Yeah, this is more to see how we'll integrate this as a POC and make sure that the tests can pass. Once this is merged, I'll add support for the rest of the recipes. 
3. **Hardcoding iterator_finished?** Yeah, we'll change this when we _actually_ move to step-based checkpointing, but right now we expect to save on the epoch boundaries so if the epoch is cut short, then we expect that the dataloader will restart its shuffling and data provided as if the iterator has finished going through all the samples. Huge, huge thanks to @ramanishsingh for helping me debug this last issue.
4. **You removed the sampler?!!** The `StatefulDataLoader` creates a batched random sampler of it's own so there's no need for us to be creating a new one in this context. Less code means my life is easier.
5. **WTH, you removed the check for max_steps is None. Won't that break?** Nope, Python is dumb. Check it:
```
>>> 1 == None
False
>>> 0 == None
False
>>> 5 == None
False
>>> 1000000000000000 == None
False
```

#### Changelog
What are the changes made in this PR?
* Import `StatefulDataloader` and replace it's usage in the `_setup_data` method
* Checkpoint the dataloader state dict
* Load the dataloader state dict if we are trying to resume from the checkpoint
* Update tests to match (yes the numbers changed b/c we are using a new random state - it's fine)

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [x] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: recipes/full_finetune_single_device.py [hunk 1]
```diff
@@ -302,11 +302,16 @@ def setup(self, cfg: DictConfig) -> None:
...
         # setup after both of these are initialized
         collate_name = cfg.get("collate_fn", "torchtune.data.padded_collate_sft")
-        self._sampler, self._dataloader = self._setup_data(
+        self._dataloader = self._setup_data(
             cfg_dataset=cfg.dataset,
             shuffle=cfg.shuffle,
             batch_size=cfg.batch_size,
             collate_fn=collate_name,
+            dataloader_state_dict=(
+                ckpt_dict[training.DATALOADER_KEY]
```
Comment (RdoubleA) [inline recipes/full_finetune_single_device.py:311]: should check if this key even exists for BC

File: tests/recipes/test_full_finetune_single_device.py [hunk 2]
```diff
@@ -56,7 +56,7 @@ def _get_test_config_overrides(self):
 
     def _fetch_expected_loss_values(self, model_type):
         loss_values_map = {
-            "llama2": [10.5201, 10.5217, 10.4945, 10.5136],
+            "llama2": [10.5130, 10.5215, 10.5764, 10.5046],
```
Comment (RdoubleA) [inline tests/recipes/test_full_finetune_single_device.py:59]: is this because the sampler is different??

File: recipes/full_finetune_single_device.py [hunk 3]
```diff
@@ -570,19 +576,11 @@ def _setup_data(
             raise RuntimeError("left_pad_sequence collator is only for inference.")
         collate_fn = _get_component_from_path(collate_fn)
 
-        sampler = DistributedSampler(
-            ds,
-            num_replicas=1,
-            rank=0,
-            shuffle=shuffle,
-            seed=0,
```
Comment (RdoubleA) [inline recipes/full_finetune_single_device.py:578]: so how is seed passed to StatefulDataLoader in this case?

File: recipes/full_finetune_single_device.py (hunk 1 above)
Comment (joecummings) [reply recipes/full_finetune_single_device.py:311]: This might just have to break BC b/c without this, the user will not be able to successfully resume training at any point.

**Review by SalmanMohammadi — APPROVED** [review]
